File: src/lib/library/v2/quantize.ts
import { GpuDesignerNode } from "../../designer/gpudesignernode";

export class Quantize extends GpuDesignerNode {
	public init() {
		this.title = "Quantize";

		this.addInput("image");

		this.addIntProperty("steps", "Steps", 12, 2, 256, 1);

		const source = `
        vec4 process(vec2 uv)
        {
            vec4 a =  texture(image, uv);
            a.rgb = floor(a.rgb * vec3(float(prop_steps))) / vec3(float(prop_steps));

            return a;
        }
        `;

		this.buildShader(source);
	}
}

File: src/texturelab/libraries/v2/quantize.cpp
#include "../../models.h"
#include "../../props.h"
#include "../libv2.h"

void QuantizeNode::init()
{
    this->title = "Quantize";

    this->addInput("image");

    this->addIntProp("steps", "Steps", 12, 2, 20, 1);

    auto source = R""""(
        vec4 process(vec2 uv)
        {
            vec4 a =  texture(image, uv);
            a.rgb = floor(a.rgb * vec3(float(prop_steps))) / vec3(float(prop_steps));

            return a;
        }
        )"""";

    this->setShaderSource(source);
}
